# inicio

In [1]:
#solo es para ver si esta tomando mi ambiente
import sys
print(sys.executable)

c:\proye_shada\Comercio\.venv\Scripts\python.exe


In [2]:
# Cargar todas las librerías
import pandas as pd
import os       #para ver si logro llegar al archivo con la ruta de datos

In [3]:
print("Carpeta actual:", os.getcwd())
print("Contenido de la carpeta:", os.listdir())

#C:\proye_shada\Comercio\data\Online_Retail.xlsx
#df_retail = pd.read_excel('data/Online_Retail.xlsx')---------Esta ruta es para python
df_retail = pd.read_excel('../data/Online_Retail.xlsx')

Carpeta actual: c:\proye_shada\Comercio\notebooks
Contenido de la carpeta: ['comercio.ipynb', 'comercio.py', 'comercio2.ipynb']


In [4]:
# Muestra los nombres de las columnas

print(df_retail.columns)

Index(['InvoiceNo', 'StockCode', 'Description', 'Quantity', 'InvoiceDate',
       'UnitPrice', 'CustomerID', 'Country'],
      dtype='str')


In [5]:
#se procede a cambiar los nombres de las columnas por unas mas claras
#cambio de nombres de las columnas

nuevos_nombres = {
    'InvoiceNo': 'num_factura',
    'StockCode': 'cod_articulo',
    'Description': 'descripcion',
    'Quantity': 'cantidad',
    'InvoiceDate': 'fh_factura',
    'UnitPrice': 'precio_unitario',
    'CustomerID': 'id_cliente',
    'Country': 'pais'
}

df_retail = df_retail.rename(columns=nuevos_nombres)
print('*' * 120)

print(df_retail.columns.tolist())

************************************************************************************************************************
['num_factura', 'cod_articulo', 'descripcion', 'cantidad', 'fh_factura', 'precio_unitario', 'id_cliente', 'pais']


#### Extracion de solo 5 registros a manera de exploración

In [6]:
print('*' * 120)
print('*' * 40 +'df_retail------- solo los 5 registros' + '*' * 40)
print(df_retail.head())
print('*' * 120)

************************************************************************************************************************
****************************************df_retail------- solo los 5 registros****************************************
  num_factura cod_articulo                          descripcion  cantidad  \
0      536365       85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1      536365        71053                  WHITE METAL LANTERN         6   
2      536365       84406B       CREAM CUPID HEARTS COAT HANGER         8   
3      536365       84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4      536365       84029E       RED WOOLLY HOTTIE WHITE HEART.         6   

           fh_factura  precio_unitario  id_cliente            pais  
0 2010-12-01 08:26:00             2.55     17850.0  United Kingdom  
1 2010-12-01 08:26:00             3.39     17850.0  United Kingdom  
2 2010-12-01 08:26:00             2.75     17850.0  United Kingdom  
3 2010-12-01 08:26:00 

#### información de las columnas

In [7]:
print('*' * 120)
print('*' * 40 +'df_retail' + '*' * 40)
#display(df_calls.head())
print('*' * 120)
print(df_retail.info())
print('*' * 120)

************************************************************************************************************************
****************************************df_retail****************************************
************************************************************************************************************************
<class 'pandas.DataFrame'>
RangeIndex: 541909 entries, 0 to 541908
Data columns (total 8 columns):
 #   Column           Non-Null Count   Dtype         
---  ------           --------------   -----         
 0   num_factura      541909 non-null  object        
 1   cod_articulo     541909 non-null  object        
 2   descripcion      540455 non-null  object        
 3   cantidad         541909 non-null  int64         
 4   fh_factura       541909 non-null  datetime64[us]
 5   precio_unitario  541909 non-null  float64       
 6   id_cliente       406829 non-null  float64       
 7   pais             541909 non-null  str           
dtypes: datetime64[us](1), fl

#### Exploración de los valores ausentes

In [8]:
#antes de comenzar otra vez con cambios hagamos la copia y trabajemos en ella

df_retail_2 = df_retail.copy()

In [9]:
print('*' * 120)
print('*' * 40 +'df_retail_2------- revisión de valores ausentes' + '*' * 40)
print(df_retail_2.isna().sum())
print('*' * 120)


************************************************************************************************************************
****************************************df_retail_2------- revisión de valores ausentes****************************************
num_factura             0
cod_articulo            0
descripcion          1454
cantidad                0
fh_factura              0
precio_unitario         0
id_cliente         135080
pais                    0
dtype: int64
************************************************************************************************************************


📌 Se observan que hay valores ausentes en la descripción y en el id_cliente, antes de proceder a eliminarlos , vamos a revisar si hay valores duplicados, para realizar una limpieza mas generica creo que lo mas importante es quitar duplicados, por que eso no afectaria mucho que si elimino algunos datos con valores con inconsitencias


Tomemos un punto importante

En análisis de datos, limpiar no significa eliminar todo lo que parece extraño.

Por ejemplo:

* Una cantidad negativa podría ser una devolución.

* Un precio cero podría corresponder a una operación especial.

* Un cliente ausente no significa necesariamente que la venta sea inválida.

* Una descripción ausente sí es un problema de calidad, pero su impacto depende del análisis que quieras realizar.

Revisar si hay valores duplicados¶

In [10]:
#antes que nada vamos a cambiar el tipo flotante de id_cliente
df_retail_2['id_cliente'] = df_retail_2['id_cliente'].astype('Int64')

In [11]:
# Contar duplicados explícitos
print(df_retail_2.duplicated().sum())

5268


Paso 1. Visualiza los registros duplicados completos.antes de decidir si los quitamos o los mantenemos

In [15]:
duplicados = df_retail_2[
    df_retail_2.duplicated(keep=False)
]

duplicados.head(10)

,num_factura,cod_articulo,descripcion,cantidad,fh_factura,precio_unitario,id_cliente,pais
485,536409,22111,SCOTTIE DOG HOT WATER BOTTLE,1,2010-12-01 11:45:00,4.95,17908,United Kingdom
489,536409,22866,HAND WARMER SCOTTY DOG DESIGN,1,2010-12-01 11:45:00,2.10,17908,United Kingdom
494,536409,21866,UNION JACK FLAG LUGGAGE TAG,1,2010-12-01 11:45:00,1.25,17908,United Kingdom
517,536409,21866,UNION JACK FLAG LUGGAGE TAG,1,2010-12-01 11:45:00,1.25,17908,United Kingdom
521,536409,22900,SET 2 TEA TOWELS I LOVE LONDON,1,2010-12-01 11:45:00,2.95,17908,United Kingdom
527,536409,22866,HAND WARMER SCOTTY DOG DESIGN,1,2010-12-01 11:45:00,2.10,17908,United Kingdom
537,536409,22900,SET 2 TEA TOWELS I LOVE LONDON,1,2010-12-01 11:45:00,2.95,17908,United Kingdom
539,536409,22111,SCOTTIE DOG HOT WATER BOTTLE,1,2010-12-01 11:45:00,4.95,17908,United Kingdom
548,536412,22327,ROUND SNACK BOXES SET OF 4 SKULLS,1,2010-12-01 11:49:00,2.95,17920,United Kingdom
555,536412,22327,ROUND SNACK BOXES SET OF 4 SKULLS,1,2010-12-01 11:49:00,2.95,17920,United Kingdom


In [16]:
print(
    "Facturas diferentes entre los duplicados:",
    duplicados['num_factura'].nunique()
)

Facturas diferentes entre los duplicados: 1933


In [17]:
duplicados.head(10)

,num_factura,cod_articulo,descripcion,cantidad,fh_factura,precio_unitario,id_cliente,pais
485,536409,22111,SCOTTIE DOG HOT WATER BOTTLE,1,2010-12-01 11:45:00,4.95,17908,United Kingdom
489,536409,22866,HAND WARMER SCOTTY DOG DESIGN,1,2010-12-01 11:45:00,2.10,17908,United Kingdom
494,536409,21866,UNION JACK FLAG LUGGAGE TAG,1,2010-12-01 11:45:00,1.25,17908,United Kingdom
517,536409,21866,UNION JACK FLAG LUGGAGE TAG,1,2010-12-01 11:45:00,1.25,17908,United Kingdom
521,536409,22900,SET 2 TEA TOWELS I LOVE LONDON,1,2010-12-01 11:45:00,2.95,17908,United Kingdom
527,536409,22866,HAND WARMER SCOTTY DOG DESIGN,1,2010-12-01 11:45:00,2.10,17908,United Kingdom
537,536409,22900,SET 2 TEA TOWELS I LOVE LONDON,1,2010-12-01 11:45:00,2.95,17908,United Kingdom
539,536409,22111,SCOTTIE DOG HOT WATER BOTTLE,1,2010-12-01 11:45:00,4.95,17908,United Kingdom
548,536412,22327,ROUND SNACK BOXES SET OF 4 SKULLS,1,2010-12-01 11:49:00,2.95,17920,United Kingdom
555,536412,22327,ROUND SNACK BOXES SET OF 4 SKULLS,1,2010-12-01 11:49:00,2.95,17920,United Kingdom


In [18]:
df_retail_2[
    df_retail_2.duplicated(keep=False)
].sort_values(
    by=['num_factura', 'cod_articulo']
).head(20)


,num_factura,cod_articulo,descripcion,cantidad,fh_factura,precio_unitario,id_cliente,pais
494,536409,21866,UNION JACK FLAG LUGGAGE TAG,1,2010-12-01 11:45:00,1.25,17908,United Kingdom
517,536409,21866,UNION JACK FLAG LUGGAGE TAG,1,2010-12-01 11:45:00,1.25,17908,United Kingdom
485,536409,22111,SCOTTIE DOG HOT WATER BOTTLE,1,2010-12-01 11:45:00,4.95,17908,United Kingdom
539,536409,22111,SCOTTIE DOG HOT WATER BOTTLE,1,2010-12-01 11:45:00,4.95,17908,United Kingdom
489,536409,22866,HAND WARMER SCOTTY DOG DESIGN,1,2010-12-01 11:45:00,2.10,17908,United Kingdom
527,536409,22866,HAND WARMER SCOTTY DOG DESIGN,1,2010-12-01 11:45:00,2.10,17908,United Kingdom
521,536409,22900,SET 2 TEA TOWELS I LOVE LONDON,1,2010-12-01 11:45:00,2.95,17908,United Kingdom
537,536409,22900,SET 2 TEA TOWELS I LOVE LONDON,1,2010-12-01 11:45:00,2.95,17908,United Kingdom
565,536412,21448,12 DAISY PEGS IN WOOD BOX,2,2010-12-01 11:49:00,1.65,17920,United Kingdom
578,536412,21448,12 DAISY PEGS IN WOOD BOX,1,2010-12-01 11:49:00,1.65,17920,United Kingdom


In [19]:
df_retail_2[
    (df_retail_2['num_factura'] == '536409') &
    (df_retail_2['cod_articulo'].isin(['22111', '22866', '22900']))
]

,num_factura,cod_articulo,descripcion,cantidad,fh_factura,precio_unitario,id_cliente,pais


In [20]:
df_retail_2[
    (df_retail_2['num_factura'] == '536409') &
    df_retail_2.duplicated(keep=False)
]

,num_factura,cod_articulo,descripcion,cantidad,fh_factura,precio_unitario,id_cliente,pais


In [21]:
print(df_retail_2['num_factura'].dtype)
print(df_retail_2['cod_articulo'].dtype)

object
object


In [22]:
df_retail_2[
    df_retail_2['num_factura'].astype(str).str.strip() == '536409'
]

,num_factura,cod_articulo,descripcion,cantidad,fh_factura,precio_unitario,id_cliente,pais
483,536409,90199C,5 STRAND GLASS NECKLACE CRYSTAL,3,2010-12-01 11:45:00,6.35,17908,United Kingdom
484,536409,21479,WHITE SKULL HOT WATER BOTTLE,1,2010-12-01 11:45:00,3.75,17908,United Kingdom
485,536409,22111,SCOTTIE DOG HOT WATER BOTTLE,1,2010-12-01 11:45:00,4.95,17908,United Kingdom
486,536409,22785,SQUARECUSHION COVER PINK UNION FLAG,1,2010-12-01 11:45:00,6.75,17908,United Kingdom
487,536409,22975,SPACEBOY CHILDRENS EGG CUP,1,2010-12-01 11:45:00,1.25,17908,United Kingdom
488,536409,22972,CHILDREN'S SPACEBOY MUG,1,2010-12-01 11:45:00,1.65,17908,United Kingdom
489,536409,22866,HAND WARMER SCOTTY DOG DESIGN,1,2010-12-01 11:45:00,2.10,17908,United Kingdom
490,536409,22568,FELTCRAFT CUSHION OWL,1,2010-12-01 11:45:00,3.75,17908,United Kingdom
491,536409,85116,BLACK CANDELABRA T-LIGHT HOLDER,1,2010-12-01 11:45:00,2.10,17908,United Kingdom
492,536409,22664,TOY TIDY DOLLY GIRL DESIGN,1,2010-12-01 11:45:00,2.10,17908,United Kingdom


In [23]:
factura_536409 = df_retail_2[
    df_retail_2['num_factura'].astype(str).str.strip() == '536409'
]

factura_536409[
    factura_536409.duplicated(keep=False)
].sort_values('cod_articulo')

,num_factura,cod_articulo,descripcion,cantidad,fh_factura,precio_unitario,id_cliente,pais
494,536409,21866,UNION JACK FLAG LUGGAGE TAG,1,2010-12-01 11:45:00,1.25,17908,United Kingdom
517,536409,21866,UNION JACK FLAG LUGGAGE TAG,1,2010-12-01 11:45:00,1.25,17908,United Kingdom
485,536409,22111,SCOTTIE DOG HOT WATER BOTTLE,1,2010-12-01 11:45:00,4.95,17908,United Kingdom
539,536409,22111,SCOTTIE DOG HOT WATER BOTTLE,1,2010-12-01 11:45:00,4.95,17908,United Kingdom
489,536409,22866,HAND WARMER SCOTTY DOG DESIGN,1,2010-12-01 11:45:00,2.10,17908,United Kingdom
527,536409,22866,HAND WARMER SCOTTY DOG DESIGN,1,2010-12-01 11:45:00,2.10,17908,United Kingdom
521,536409,22900,SET 2 TEA TOWELS I LOVE LONDON,1,2010-12-01 11:45:00,2.95,17908,United Kingdom
537,536409,22900,SET 2 TEA TOWELS I LOVE LONDON,1,2010-12-01 11:45:00,2.95,17908,United Kingdom


In [24]:
# Contar los registros duplicados exactos
duplicados_exactos = df_retail_2.duplicated().sum()

print("Registros duplicados exactos:", duplicados_exactos)

Registros duplicados exactos: 5268


In [25]:
# antes de eliminar hagamos un respaldo hasta aqui

df_retail_3 = df_retail_2.copy()

In [26]:
df_retail_sin_duplicados = df_retail_3.drop_duplicates()

print("Filas originales:", len(df_retail_3))
print("Filas sin duplicados:", len(df_retail_sin_duplicados))
print("Filas eliminadas:", len(df_retail_3) - len(df_retail_sin_duplicados))

Filas originales: 541909
Filas sin duplicados: 536641
Filas eliminadas: 5268


In [27]:
print('*' * 120)
print('*' * 40 +'df_retail_sin_duplicados------- revisión de valores ausentes' + '*' * 40)
print(df_retail_sin_duplicados.isna().sum())
print('*' * 120)


************************************************************************************************************************
****************************************df_retail_sin_duplicados------- revisión de valores ausentes****************************************
num_factura             0
cod_articulo            0
descripcion          1454
cantidad                0
fh_factura              0
precio_unitario         0
id_cliente         135037
pais                    0
dtype: int64
************************************************************************************************************************


In [28]:
porcentaje_ausentes = (
    df_retail_sin_duplicados.isna().mean() * 100
).round(2)

print(porcentaje_ausentes.sort_values(ascending=False))

id_cliente         25.16
descripcion         0.27
cod_articulo        0.00
num_factura         0.00
cantidad            0.00
fh_factura          0.00
precio_unitario     0.00
pais                0.00
dtype: float64


Aqui se ve claramnete como se comportan los valores ausentes:

* los ---->id_cliente tiene un 25.16 %----->Aproximadamente 1 de cada 4 registros no tiene identificado al cliente.
* ----->descripcion tienen un 0.27 % ------->Es una proporción pequeña; posiblemente podamos recuperar algunas descripciones.
* ----->las demas columnas tiene un 0%------>no presentan valores ausentes.

In [29]:
#revisemos los registros que no tienen descripción
df_retail_sin_duplicados[
    df_retail_sin_duplicados['descripcion'].isna()
][['cod_articulo', 'descripcion', 'cantidad', 'precio_unitario']]

,cod_articulo,descripcion,cantidad,precio_unitario
622,22139,NaN,56,0.0
1970,21134,NaN,1,0.0
1971,22145,NaN,1,0.0
1972,37509,NaN,1,0.0
1987,85226A,NaN,1,0.0
...,...,...,...,...
535322,84581,NaN,-2,0.0
535326,23406,NaN,15,0.0
535332,21620,NaN,6,0.0
536981,72817,NaN,27,0.0


Hay dos cosas interesantes en lo que muestras:

Algunos artículos tienen cantidades negativas, como -2.

Todos los ejemplos visibles tienen precio_unitario = 0.0.

In [30]:
#Comprobar si podemos recuperar las descripciones

descripciones_por_codigo = (
    df_retail_sin_duplicados
    .dropna(subset=['descripcion'])
    .groupby('cod_articulo')['descripcion']
    .nunique()
)

codigos_con_varias_descripciones = (
    descripciones_por_codigo[descripciones_por_codigo > 1]
)

print(
    "Códigos con varias descripciones:",
    len(codigos_con_varias_descripciones)
)

Códigos con varias descripciones: 650


In [31]:
#Comprobar cuántas descripciones podemos recuperar
# Obtener los registros donde falta la descripción
sin_descripcion = df_retail_sin_duplicados[
    df_retail_sin_duplicados['descripcion'].isna()
]

# Contar cuántos códigos faltantes tienen una única descripción conocida
codigos_unicos = descripciones_por_codigo[
    descripciones_por_codigo == 1
].index

recuperables = sin_descripcion[
    sin_descripcion['cod_articulo'].isin(codigos_unicos)
]

print("Registros sin descripción:", len(sin_descripcion))
print("Registros candidatos a recuperar:", len(recuperables))
print(
    "Registros que requieren más revisión:",
    len(sin_descripcion) - len(recuperables)
)



Registros sin descripción: 1454
Registros candidatos a recuperar: 1033
Registros que requieren más revisión: 421


* Registros sin descripción          ----->1,454

* 1,033 candidatos a recuperar------->Tienen un código con una sola descripción conocida.------>71.05 %

* 421 registros por investigar-------->No cumplen el criterio de tener una única descripción conocida.------->28.95 %

In [32]:
recuperables[
    ['cod_articulo', 'descripcion', 'cantidad', 'precio_unitario']
].head(10)

,cod_articulo,descripcion,cantidad,precio_unitario
1971,22145,NaN,1,0.0
1972,37509,NaN,1,0.0
2025,37461,NaN,3,0.0
2406,21777,NaN,-10,0.0
4347,84952C,NaN,-38,0.0
4348,84952C,NaN,19,0.0
7188,22712,NaN,-20,0.0
7189,22028,NaN,-20,0.0
7190,85067,NaN,-6,0.0
7192,21414,NaN,-22,0.0


In [33]:
mapa_descripciones = (
    df_retail_sin_duplicados
    .dropna(subset=['descripcion'])
    .groupby('cod_articulo')['descripcion']
    .first()
)

recuperables_con_descripcion = recuperables[
    ['cod_articulo']
].copy()

recuperables_con_descripcion['descripcion_propuesta'] = (
    recuperables_con_descripcion['cod_articulo']
    .map(mapa_descripciones)
)

recuperables_con_descripcion.head(10)

,cod_articulo,descripcion_propuesta
1971,22145,CHRISTMAS CRAFT HEART STOCKING
1972,37509,NEW ENGLAND MUG W GIFT BOX
2025,37461,FUNKY MONKEY MUG
2406,21777,RECIPE BOX WITH METAL HEART
4347,84952C,MIRROR LOVE BIRD T-LIGHT HOLDER
4348,84952C,MIRROR LOVE BIRD T-LIGHT HOLDER
7188,22712,CARD DOLLY GIRL
7189,22028,PENNY FARTHING BIRTHDAY CARD
7190,85067,CREAM SWEETHEART WALL CABINET
7192,21414,SCALLOP SHELL SOAP DISH


In [34]:
# Comprobar que cada código candidato tenga una descripción propuesta
print(
    "Candidatos:",
    len(recuperables_con_descripcion)
)

print(
    "Candidatos sin descripción propuesta:",
    recuperables_con_descripcion['descripcion_propuesta'].isna().sum()
)

print(
    "Códigos candidatos con una sola descripción conocida:",
    recuperables['cod_articulo'].isin(codigos_unicos).sum()
)

Candidatos: 1033
Candidatos sin descripción propuesta: 0
Códigos candidatos con una sola descripción conocida: 1033


In [35]:
df_retail_limpio = df_retail_sin_duplicados.copy()

print("Filas de la copia:", len(df_retail_limpio))
print(
    "Descripciones ausentes:",
    df_retail_limpio['descripcion'].isna().sum()
)

Filas de la copia: 536641
Descripciones ausentes: 1454


In [36]:
descripciones_por_codigo = (
    df_retail_limpio
    .dropna(subset=['descripcion'])
    .groupby('cod_articulo')['descripcion']
    .nunique()
)

codigos_con_descripcion_unica = descripciones_por_codigo[
    descripciones_por_codigo == 1
].index

print(
    "Códigos con una sola descripción conocida:",
    len(codigos_con_descripcion_unica)
)

Códigos con una sola descripción conocida: 3308


In [37]:
# Seleccionar los registros donde falta la descripción
sin_descripcion = df_retail_limpio[
    df_retail_limpio['descripcion'].isna()
]

# Identificar los registros cuyos códigos tienen una sola descripción conocida
recuperables = sin_descripcion[
    sin_descripcion['cod_articulo'].isin(
        codigos_con_descripcion_unica
    )
]

print("Registros sin descripción:", len(sin_descripcion))
print("Candidatos a recuperar:", len(recuperables))
print(
    "Registros pendientes de revisión:",
    len(sin_descripcion) - len(recuperables)
)

Registros sin descripción: 1454
Candidatos a recuperar: 1033
Registros pendientes de revisión: 421


In [38]:
# Crear un mapa de códigos con una única descripción conocida
mapa_descripciones = (
    df_retail_limpio
    .dropna(subset=['descripcion'])
    .groupby('cod_articulo')['descripcion']
    .first()
)

# Crear una tabla de revisión para los candidatos
recuperables_con_descripcion = recuperables[
    ['cod_articulo']
].copy()

recuperables_con_descripcion['descripcion_propuesta'] = (
    recuperables_con_descripcion['cod_articulo']
    .map(mapa_descripciones)
)

# Mostrar algunos ejemplos
print(recuperables_con_descripcion.head(10))

     cod_articulo            descripcion_propuesta
1971        22145  CHRISTMAS CRAFT HEART STOCKING 
1972        37509       NEW ENGLAND MUG W GIFT BOX
2025        37461                 FUNKY MONKEY MUG
2406        21777      RECIPE BOX WITH METAL HEART
4347       84952C  MIRROR LOVE BIRD T-LIGHT HOLDER
4348       84952C  MIRROR LOVE BIRD T-LIGHT HOLDER
7188        22712                 CARD DOLLY GIRL 
7189        22028     PENNY FARTHING BIRTHDAY CARD
7190        85067    CREAM SWEETHEART WALL CABINET
7192        21414          SCALLOP SHELL SOAP DISH


In [40]:
# Completar las descripciones verificadas

# Identificar los registros sin descripción
# cuyo código tiene una única descripción conocida
mascara = (
    df_retail_limpio['descripcion'].isna()
    & df_retail_limpio['cod_articulo'].isin(
        codigos_con_descripcion_unica
    )
)

# Completar solo esos registros
df_retail_limpio.loc[mascara, 'descripcion'] = (
    df_retail_limpio.loc[mascara, 'cod_articulo']
    .map(mapa_descripciones)
)

# Comprobar los resultados
print(
    "Descripciones ausentes restantes:",
    df_retail_limpio['descripcion'].isna().sum()
)

print("Filas totales:", len(df_retail_limpio))



Descripciones ausentes restantes: 421
Filas totales: 536641


In [41]:
pendientes_descripcion = df_retail_limpio[
    df_retail_limpio['descripcion'].isna()
]

print("Registros pendientes:", len(pendientes_descripcion))

print(
    pendientes_descripcion[
        ['cod_articulo', 'cantidad', 'precio_unitario']
    ].head(15)
)

Registros pendientes: 421
      cod_articulo  cantidad  precio_unitario
622          22139        56              0.0
1970         21134         1              0.0
1987        85226A         1              0.0
1988         85044         1              0.0
2024         20950         1              0.0
2026         84670        23              0.0
7187         35951        57              0.0
7191         21421       110              0.0
7193         21653        -6              0.0
7203         22139        -5              0.0
7205         35965       -25              0.0
7209        47556B       -69              0.0
13263        85215       120              0.0
13961        22959        13              0.0
14065        21349       -44              0.0


In [42]:
# Revisar los códigos de los registros pendientes
codigos_pendientes = pendientes_descripcion['cod_articulo'].unique()

# Contar cuántos tienen varias descripciones conocidas
con_varias = [
    codigo for codigo in codigos_pendientes
    if descripciones_por_codigo.get(codigo, 0) > 1
]

# Contar cuántos tienen una descripción conocida
con_una = [
    codigo for codigo in codigos_pendientes
    if descripciones_por_codigo.get(codigo, 0) == 1
]

# Contar cuántos no tienen ninguna descripción conocida
sin_ninguna = [
    codigo for codigo in codigos_pendientes
    if descripciones_por_codigo.get(codigo, 0) == 0
]

print("Códigos pendientes con varias descripciones:", len(con_varias))
print("Códigos pendientes con una descripción conocida:", len(con_una))
print("Códigos pendientes sin descripción conocida:", len(sin_ninguna))

Códigos pendientes con varias descripciones: 174
Códigos pendientes con una descripción conocida: 0
Códigos pendientes sin descripción conocida: 112


In [43]:
# Contar cuántas descripciones conocidas tiene cada código pendiente
cantidad_descripciones = (
    pendientes_descripcion['cod_articulo']
    .map(descripciones_por_codigo)
    .fillna(0)
)

print(
    "Registros cuyos códigos tienen varias descripciones:",
    (cantidad_descripciones > 1).sum()
)

print(
    "Registros cuyos códigos no tienen descripción conocida:",
    (cantidad_descripciones == 0).sum()
)

print(
    "Total de registros pendientes:",
    len(pendientes_descripcion)
)

Registros cuyos códigos tienen varias descripciones: 309
Registros cuyos códigos no tienen descripción conocida: 112
Total de registros pendientes: 421


In [45]:
# Seleccionar algunos códigos pendientes con varias descripciones
codigos_revision = con_varias[:10]

# Obtener sus descripciones conocidas
ejemplos_descripciones = (
    df_retail_sin_duplicados[
        df_retail_sin_duplicados['cod_articulo'].isin(codigos_revision)
        & df_retail_sin_duplicados['descripcion'].notna()
    ][['cod_articulo', 'descripcion']]
    .drop_duplicates()
    .copy()
)

# Ordenar los códigos como texto, sin modificar los datos originales
ejemplos_descripciones = ejemplos_descripciones.sort_values(
    'cod_articulo',
    key=lambda columna: columna.astype(str)
)

print(ejemplos_descripciones.to_string(index=False))

cod_articulo                         descripcion
      17091A             LAVENDER INCENSE IN TIN
      17091A                               found
       20956  PORCELAIN T-LIGHT HOLDERS ASSORTED
       20956                               check
       20956                                   ?
       21349                    IVY HEART WREATH
       21349                               check
       21421               PORCELAIN ROSE LARGE 
       21421                             damaged
       22139    RETROSPOT TEA SET CERAMIC 11 PC 
       22139                              amazon
       22959                               check
       22959              WRAP CHRISTMAS VILLAGE
       35965            found some more on shelf
       35965                               found
       35965                              dotcom
       35965          FOLKART HEART NAPKIN RINGS
      47556B did  a credit  and did not tick ret
      47556B                TEA TIME TEA TOWELS 
      85167B        

In [46]:
print("Registros finales:", len(df_retail_limpio))
print("Descripciones faltantes:", df_retail_limpio['descripcion'].isna().sum())
print("Clientes sin identificar:", df_retail_limpio['id_cliente'].isna().sum())
print("Filas duplicadas exactas:", df_retail_limpio.duplicated().sum())

Registros finales: 536641
Descripciones faltantes: 421
Clientes sin identificar: 135037
Filas duplicadas exactas: 0


In [47]:
print("Cantidades negativas:", (df_retail_limpio['cantidad'] < 0).sum())
print("Precios negativos:", (df_retail_limpio['precio_unitario'] < 0).sum())
print("Precios iguales a cero:", (df_retail_limpio['precio_unitario'] == 0).sum())

display(
    df_retail_limpio[
        (df_retail_limpio['cantidad'] < 0) |
        (df_retail_limpio['precio_unitario'] < 0) |
        (df_retail_limpio['precio_unitario'] == 0)
    ][['num_factura', 'cod_articulo', 'descripcion',
       'cantidad', 'precio_unitario']].head(10)
)

Cantidades negativas: 10587
Precios negativos: 2
Precios iguales a cero: 2510


,num_factura,cod_articulo,descripcion,cantidad,precio_unitario
141,C536379,D,Discount,-1,27.50
154,C536383,35004C,SET OF 3 COLOURED FLYING DUCKS,-1,4.65
235,C536391,22556,PLASTERS IN TIN CIRCUS PARADE,-12,1.65
236,C536391,21984,PACK OF 12 PINK PAISLEY TISSUES,-24,0.29
237,C536391,21983,PACK OF 12 BLUE PAISLEY TISSUES,-24,0.29
238,C536391,21980,PACK OF 12 RED RETROSPOT TISSUES,-24,0.29
239,C536391,21484,CHICK GREY HOT WATER BOTTLE,-12,3.45
240,C536391,22557,PLASTERS IN TIN VINTAGE PAISLEY,-12,1.65
241,C536391,22553,PLASTERS IN TIN SKULLS,-24,1.65
622,536414,22139,NaN,56,0.00


In [ ]:
********************************aqui me quede hasta mañana

In [18]:
df_retail_2 = df_retail_2.drop_duplicates().copy()

In [20]:
df_retail_2[df_retail_2.duplicated(keep=False)].head(10)

,num_factura,cod_articulo,descripcion,cantidad,fh_factura,precio_unitario,id_cliente,pais


📌 Se encuentra que hay 5268 datos duplicados, por lo que procedemos a quitarlos y veamos que pasa esto lo estamos realizando dentro de la copia, no en el dataframe original

In [21]:
#comprobar si hay quedaron valores duplicados
print(df_retail_2.duplicated().sum())

0


vemos si una vez que se quitaron los valores duplicados, como salen los registros

In [ ]:
print('*' * 120)
print('*' * 40 +'df_retail_2------- revisión de valores ausentes' + '*' * 40)
print(df_retail_2.isna().sum())
print('*' * 120)


************************************************************************************************************************
****************************************df_retail------- revisión de valores ausentes****************************************
num_factura             0
cod_articulo            0
descripcion          1454
cantidad                0
fh_factura              0
precio_initario         0
id_cliente         135037
pais                    0
dtype: int64
************************************************************************************************************************


In [22]:
print('*' * 120)
print(df_retail_2.info())
print('*' * 120)

************************************************************************************************************************
<class 'pandas.DataFrame'>
Index: 536641 entries, 0 to 541908
Data columns (total 8 columns):
 #   Column           Non-Null Count   Dtype         
---  ------           --------------   -----         
 0   num_factura      536641 non-null  object        
 1   cod_articulo     536641 non-null  object        
 2   descripcion      535187 non-null  object        
 3   cantidad         536641 non-null  int64         
 4   fh_factura       536641 non-null  datetime64[us]
 5   precio_unitario  536641 non-null  float64       
 6   id_cliente       401604 non-null  Int64         
 7   pais             536641 non-null  str           
dtypes: Int64(1), datetime64[us](1), float64(1), int64(1), object(3), str(1)
memory usage: 37.4+ MB
None
************************************************************************************************************************


In [23]:
# Calcular el número de valores ausentes
print(df_retail_2.isna().sum())

num_factura             0
cod_articulo            0
descripcion          1454
cantidad                0
fh_factura              0
precio_unitario         0
id_cliente         135037
pais                    0
dtype: int64


### Si quiero ver las filas que tienen valores ausentes

In [2]:
df_retail_2[df_retail_2['descripcion'].isna()].head(10)

NameError: name 'df_retail_2' is not defined

Aqui hagamos una pausa y recordemos lo de antes se menciono 🤔🤔🤔🤔🤔🤔🤔🤔🤔🤔🤔🤔🤔🤔🤔🤔🤔🤔🤔🤔

* Una cantidad negativa podría ser una devolución.

* Un precio cero podría corresponder a una operación especial.

* Un cliente ausente no significa necesariamente que la venta sea inválida.

* Una descripción ausente sí es un problema de calidad, pero su impacto depende del análisis que quieras realizar.

vamos a ver como ir obteniendo la descripcion por la del cod_articulo


In [ ]:
#antes de ver que me este faltando preguntemos por este registro solamente: 22139

In [ ]:
descripciones_por_codigo = (
    df_retail_2[df_retail_2['descripcion'].notna()]
    .groupby('cod_articulo')['descripcion']
    .nunique()
)

codigos_con_varias_descripciones = descripciones_por_codigo[
    descripciones_por_codigo > 1
]

print(codigos_con_varias_descripciones)

cod_articulo
10080           2
10133           2
16008           2
16045           2
20622           2
               ..
90195A          2
90210D          2
DCGS0003        2
DCGS0069        2
gift_0001_20    2
Name: descripcion, Length: 650, dtype: int64


In [ ]:
# Crear un diccionario con el código y su descripción
mapa_descripciones = (
    df_retail_2.dropna(subset=['descripcion'])
    .drop_duplicates(subset=['cod_articulo'])
    .set_index('cod_articulo')['descripcion']
)

# Completar únicamente las descripciones ausentes
df_retail_2['descripcion'] = (
    df_retail_2['descripcion'].fillna(
        df_retail_2['cod_articulo'].map(mapa_descripciones)
    )
)

In [ ]:
ausentes_antes = df_retail_2['descripcion'].isna().sum()

print("Descripciones ausentes:", ausentes_antes)

Descripciones ausentes: 112


In [ ]:
print(
    "Descripciones ausentes después:",
    df_retail_2['descripcion'].isna().sum()
)

Descripciones ausentes después: 112


In [ ]:
# Calcular el número de valores ausentes
print(df_retail_2.isna().sum())

num_factura             0
cod_articulo            0
descripcion           112
cantidad                0
fh_factura              0
precio_unitario         0
id_cliente         135037
pais                    0
dtype: int64


In [ ]:
print(df_retail_2.isna().sum())

num_factura             0
cod_articulo            0
descripcion          1454
cantidad                0
fh_factura              0
precio_unitario         0
id_cliente         135037
pais                    0
dtype: int64


Aqui se esta observando algo que el precio unitario esta en 0.0, por lo que procedemos agregarle un filtro de que sea mayor a 0.0


,num_factura,cod_articulo,descripcion,cantidad,fh_factura,precio_initario,id_cliente,pais


Despues de esto observamos que no hay es necesario estos valores por lo que aunque dicen que cantidad algunos son mayores a 1 unidad por lo cual no se verea una afectacion si quitamos completamente estos registro


In [ ]:
df_retail[df_retail['precio_initario'].isna()].head(10)

,num_factura,cod_articulo,descripcion,cantidad,fh_factura,precio_initario,id_cliente,pais


In [ ]:
# hagamos un respaldo antes 
df_retail_2 = df_retail

#ahora si eliminemos esos registros que se ve que no afecta que podamos eliminarlos
df_retail_2[df_retail_2['descripcion'].isna()].head(10)

,num_factura,cod_articulo,descripcion,cantidad,fh_factura,precio_initario,id_cliente,pais
622,536414,22139,NaN,56,2010-12-01 11:52:00,0.0,<NA>,United Kingdom
1970,536545,21134,NaN,1,2010-12-01 14:32:00,0.0,<NA>,United Kingdom
1971,536546,22145,NaN,1,2010-12-01 14:33:00,0.0,<NA>,United Kingdom
1972,536547,37509,NaN,1,2010-12-01 14:33:00,0.0,<NA>,United Kingdom
1987,536549,85226A,NaN,1,2010-12-01 14:34:00,0.0,<NA>,United Kingdom
1988,536550,85044,NaN,1,2010-12-01 14:34:00,0.0,<NA>,United Kingdom
2024,536552,20950,NaN,1,2010-12-01 14:34:00,0.0,<NA>,United Kingdom
2025,536553,37461,NaN,3,2010-12-01 14:35:00,0.0,<NA>,United Kingdom
2026,536554,84670,NaN,23,2010-12-01 14:35:00,0.0,<NA>,United Kingdom
2406,536589,21777,NaN,-10,2010-12-01 16:50:00,0.0,<NA>,United Kingdom


Eliminemos los duplicados explicitamente

In [ ]:
df_retail_2.drop_duplicates(inplace=True)

In [ ]:
print(df_retail_2.duplicated().sum())

NameError: name 'df_retail_2' is not defined

Eliminemos los valores ausentes, revisemos antes si esos asusenten cumplen que sea mayor de 0.0 en precio unitario

In [ ]:
df_retail_2[df_retail_2['descripcion'].isna() &
          df_retail_2['precio_initario']>0.0].head(10)

,num_factura,cod_articulo,descripcion,cantidad,fh_factura,precio_initario,id_cliente,pais


In [ ]:
df_retail_2 = df_retail_2.drop(
    df_retail_2[
        df_retail_2['descripcion'].isna() &
        (df_retail_2['precio_initario'] == 0.0)
    ].index
)

In [ ]:
# Calcular el número de valores ausentes
print(df_retail_2.isna().sum())

num_factura             0
cod_articulo            0
descripcion             0
cantidad                0
fh_factura              0
precio_initario         0
id_cliente         133583
pais                    0
dtype: int64


In [ ]:
print(df_retail_2.head())

  num_factura cod_articulo                          descripcion  cantidad  \
0      536365       85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1      536365        71053                  WHITE METAL LANTERN         6   
2      536365       84406B       CREAM CUPID HEARTS COAT HANGER         8   
3      536365       84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4      536365       84029E       RED WOOLLY HOTTIE WHITE HEART.         6   

           fh_factura  precio_initario id_cliente            pais  
0 2010-12-01 08:26:00             2.55    17850.0  United Kingdom  
1 2010-12-01 08:26:00             3.39    17850.0  United Kingdom  
2 2010-12-01 08:26:00             2.75    17850.0  United Kingdom  
3 2010-12-01 08:26:00             3.39    17850.0  United Kingdom  
4 2010-12-01 08:26:00             3.39    17850.0  United Kingdom  


In [ ]:
df_retail_2[df_retail_2['id_cliente'].isna() &
          (df_retail_2['precio_initario']==0)].head(10)

,num_factura,cod_articulo,descripcion,cantidad,fh_factura,precio_initario,id_cliente,pais
6391,536941,22734,amazon,20,2010-12-03 12:08:00,0.0,<NA>,United Kingdom
6392,536942,22139,amazon,15,2010-12-03 12:08:00,0.0,<NA>,United Kingdom
7313,537032,21275,?,-30,2010-12-03 16:50:00,0.0,<NA>,United Kingdom
13217,537425,84968F,check,-20,2010-12-06 15:35:00,0.0,<NA>,United Kingdom
13218,537426,84968E,check,-35,2010-12-06 15:36:00,0.0,<NA>,United Kingdom
13264,537432,35833G,damages,-43,2010-12-06 16:10:00,0.0,<NA>,United Kingdom
14335,537534,85064,CREAM SWEETHEART LETTER RACK,1,2010-12-07 11:48:00,0.0,<NA>,United Kingdom
14336,537534,84832,ZINC WILLIE WINKIE CANDLE STICK,1,2010-12-07 11:48:00,0.0,<NA>,United Kingdom
14337,537534,84692,BOX OF 24 COCKTAIL PARASOLS,2,2010-12-07 11:48:00,0.0,<NA>,United Kingdom
14338,537534,48184,DOORMAT ENGLISH ROSE,3,2010-12-07 11:48:00,0.0,<NA>,United Kingdom


In [ ]:
df_retail_2[df_retail_2['id_cliente'].isna() &
            (df_retail_2['cantidad']<0) &
          (df_retail_2['precio_initario']==0)].head(10)

,num_factura,cod_articulo,descripcion,cantidad,fh_factura,precio_initario,id_cliente,pais
7313,537032,21275,?,-30,2010-12-03 16:50:00,0.0,<NA>,United Kingdom
13217,537425,84968F,check,-20,2010-12-06 15:35:00,0.0,<NA>,United Kingdom
13218,537426,84968E,check,-35,2010-12-06 15:36:00,0.0,<NA>,United Kingdom
13264,537432,35833G,damages,-43,2010-12-06 16:10:00,0.0,<NA>,United Kingdom
21338,538072,22423,faulty,-13,2010-12-09 14:10:00,0.0,<NA>,United Kingdom
21518,538090,20956,?,-723,2010-12-09 14:48:00,0.0,<NA>,United Kingdom
22296,538161,46000S,Dotcom sales,-100,2010-12-09 17:25:00,0.0,<NA>,United Kingdom
22297,538162,46000M,Dotcom sales,-100,2010-12-09 17:25:00,0.0,<NA>,United Kingdom
42564,540010,22501,reverse 21/5/10 adjustment,-100,2011-01-04 11:13:00,0.0,<NA>,United Kingdom
42566,540012,22502,reverse 21/5/10 adjustment,-100,2011-01-04 11:14:00,0.0,<NA>,United Kingdom


In [ ]:
# se observan varios registros, pero tambien quisiera saber cuantos son los registro que hay en total y para eso realizo esto

total =df_retail_2[
    df_retail_2['id_cliente'].isna() &
    (df_retail_2['cantidad']<0) &
    (df_retail_2['precio_initario']==0)]

print("cantidad que cumple esta condicion es :" , total.shape[0])

cantidad que cumple esta condicion es : 0


Viendolo de esa manera combiene mas eliminarlo y ver que queda

In [ ]:
df_retail_2 = df_retail_2.drop(
    df_retail_2[
        df_retail_2['id_cliente'].isna() &
    (df_retail_2['cantidad']<0) &
    (df_retail_2['precio_initario']==0)
    ].index
)

In [ ]:
total =df_retail_2[
    df_retail_2['id_cliente'].isna() &
    (df_retail_2['cantidad']<0) &
    (df_retail_2['precio_initario']==0)]

print("cantidad que cumple esta condicion es :" , total.shape[0])

cantidad que cumple esta condicion es : 0


In [ ]:
# Calcular el número de valores ausentes
print(df_retail_2.isna().sum())

num_factura             0
cod_articulo            0
descripcion             0
cantidad                0
fh_factura              0
precio_initario         0
id_cliente         133109
pais                    0
dtype: int64


In [ ]:
df_retail_2[df_retail_2['id_cliente'].isna() &
            
          (df_retail_2['precio_initario']==0)].head(10)

,num_factura,cod_articulo,descripcion,cantidad,fh_factura,precio_initario,id_cliente,pais
6391,536941,22734,amazon,20,2010-12-03 12:08:00,0.0,<NA>,United Kingdom
6392,536942,22139,amazon,15,2010-12-03 12:08:00,0.0,<NA>,United Kingdom
14335,537534,85064,CREAM SWEETHEART LETTER RACK,1,2010-12-07 11:48:00,0.0,<NA>,United Kingdom
14336,537534,84832,ZINC WILLIE WINKIE CANDLE STICK,1,2010-12-07 11:48:00,0.0,<NA>,United Kingdom
14337,537534,84692,BOX OF 24 COCKTAIL PARASOLS,2,2010-12-07 11:48:00,0.0,<NA>,United Kingdom
14338,537534,48184,DOORMAT ENGLISH ROSE,3,2010-12-07 11:48:00,0.0,<NA>,United Kingdom
14339,537534,48111,DOORMAT 3 SMILEY CATS,1,2010-12-07 11:48:00,0.0,<NA>,United Kingdom
14340,537534,22697,GREEN REGENCY TEACUP AND SAUCER,1,2010-12-07 11:48:00,0.0,<NA>,United Kingdom
14341,537534,22682,FRENCH BLUE METAL DOOR SIGN 7,1,2010-12-07 11:48:00,0.0,<NA>,United Kingdom
14342,537534,22680,FRENCH BLUE METAL DOOR SIGN 5,1,2010-12-07 11:48:00,0.0,<NA>,United Kingdom


In [ ]:
total =df_retail_2[
    df_retail_2['id_cliente'].isna() &
  
    (df_retail_2['precio_initario']==0)]

print("cantidad que cumple esta condicion es :" , total.shape[0])

cantidad que cumple esta condicion es : 542


In [ ]:
df_retail_2 = df_retail_2.drop(
    df_retail_2[
    df_retail_2['id_cliente'].isna() &
    (df_retail_2['precio_initario']==0)
    ].index
)

In [ ]:
#volvemos a revisar cuantos valores ausentes hay, antes de quitar esos registros habia 542, veamos ahora cuantos quedan
# Calcular el número de valores ausentes
print(df_retail_2.isna().sum())

num_factura             0
cod_articulo            0
descripcion             0
cantidad                0
fh_factura              0
precio_initario         0
id_cliente         132567
pais                    0
dtype: int64


In [ ]:
df_retail_2[df_retail_2['id_cliente'].isna()].head(10)

,num_factura,cod_articulo,descripcion,cantidad,fh_factura,precio_initario,id_cliente,pais
1443,536544,21773,DECORATIVE ROSE BATHROOM BOTTLE,1,2010-12-01 14:32:00,2.51,<NA>,United Kingdom
1444,536544,21774,DECORATIVE CATS BATHROOM BOTTLE,2,2010-12-01 14:32:00,2.51,<NA>,United Kingdom
1445,536544,21786,POLKADOT RAIN HAT,4,2010-12-01 14:32:00,0.85,<NA>,United Kingdom
1446,536544,21787,RAIN PONCHO RETROSPOT,2,2010-12-01 14:32:00,1.66,<NA>,United Kingdom
1447,536544,21790,VINTAGE SNAP CARDS,9,2010-12-01 14:32:00,1.66,<NA>,United Kingdom
1448,536544,21791,VINTAGE HEADS AND TAILS CARD GAME,2,2010-12-01 14:32:00,2.51,<NA>,United Kingdom
1449,536544,21801,CHRISTMAS TREE DECORATION WITH BELL,10,2010-12-01 14:32:00,0.43,<NA>,United Kingdom
1450,536544,21802,CHRISTMAS TREE HEART DECORATION,9,2010-12-01 14:32:00,0.43,<NA>,United Kingdom
1451,536544,21803,CHRISTMAS TREE STAR DECORATION,11,2010-12-01 14:32:00,0.43,<NA>,United Kingdom
1452,536544,21809,CHRISTMAS HANGING TREE WITH BELL,1,2010-12-01 14:32:00,2.51,<NA>,United Kingdom


In [ ]:
df_retail_2[
    df_retail_2['id_cliente'].isna() &
  
    (df_retail_2['precio_initario']==0)]

,num_factura,cod_articulo,descripcion,cantidad,fh_factura,precio_initario,id_cliente,pais


In [ ]:
total =df_retail_2[
    df_retail_2['id_cliente'].isna() &
  
    (df_retail_2['cantidad']==-1)
    ].sort_values(
    by='precio_initario',
    ascending=False)

#print("cantidad que cumple esta condicion es :" , total.shape[0])
#print("cantidad que cumple esta condicion es :" , total)
print(total.head(10))
#print(total.sum())

print("cantidad que cumple esta condicion es :" , total.shape[0])


Empty DataFrame
Columns: [num_factura, cod_articulo, descripcion, cantidad, fh_factura, precio_initario, id_cliente, pais]
Index: []
cantidad que cumple esta condicion es : 0


In [ ]:
#se eliminaron los registros donde la cantidad es -1 por que no tiene sentido dejar este tipo de registro por mucho que tenga un gran numero para el precio unitario

df_retail_2 = df_retail_2.drop(
    df_retail_2[
    df_retail_2['id_cliente'].isna() &
    (df_retail_2['cantidad']==-1)
    ].index
)

In [ ]:
print(df_retail_2.isna().sum())

num_factura             0
cod_articulo            0
descripcion             0
cantidad                0
fh_factura              0
precio_initario         0
id_cliente         132311
pais                    0
dtype: int64


### Detangamos todo por que podemos dar otro enfoque

In [ ]:
print("DataFrame original:", df_retail.shape)
print("DataFrame de trabajo:", df_retail_2.shape)

DataFrame original: (536641, 8)
DataFrame de trabajo: (533915, 8)


In [ ]:
df_retail_2[df_retail_2['id_cliente'].isna() &
          df_retail_2['precio_initario']!=0].head(10)

,num_factura,cod_articulo,descripcion,cantidad,fh_factura,precio_initario,id_cliente,pais
1443,536544,21773,DECORATIVE ROSE BATHROOM BOTTLE,1,2010-12-01 14:32:00,2.51,<NA>,United Kingdom
1444,536544,21774,DECORATIVE CATS BATHROOM BOTTLE,2,2010-12-01 14:32:00,2.51,<NA>,United Kingdom
1445,536544,21786,POLKADOT RAIN HAT,4,2010-12-01 14:32:00,0.85,<NA>,United Kingdom
1446,536544,21787,RAIN PONCHO RETROSPOT,2,2010-12-01 14:32:00,1.66,<NA>,United Kingdom
1447,536544,21790,VINTAGE SNAP CARDS,9,2010-12-01 14:32:00,1.66,<NA>,United Kingdom
1448,536544,21791,VINTAGE HEADS AND TAILS CARD GAME,2,2010-12-01 14:32:00,2.51,<NA>,United Kingdom
1449,536544,21801,CHRISTMAS TREE DECORATION WITH BELL,10,2010-12-01 14:32:00,0.43,<NA>,United Kingdom
1450,536544,21802,CHRISTMAS TREE HEART DECORATION,9,2010-12-01 14:32:00,0.43,<NA>,United Kingdom
1451,536544,21803,CHRISTMAS TREE STAR DECORATION,11,2010-12-01 14:32:00,0.43,<NA>,United Kingdom
1452,536544,21809,CHRISTMAS HANGING TREE WITH BELL,1,2010-12-01 14:32:00,2.51,<NA>,United Kingdom


In [ ]:
df_retail_2[df_retail_2['id_cliente'].isna()].head(10)

,num_factura,cod_articulo,descripcion,cantidad,fh_factura,precio_initario,id_cliente,pais
1443,536544,21773,DECORATIVE ROSE BATHROOM BOTTLE,1,2010-12-01 14:32:00,2.51,<NA>,United Kingdom
1444,536544,21774,DECORATIVE CATS BATHROOM BOTTLE,2,2010-12-01 14:32:00,2.51,<NA>,United Kingdom
1445,536544,21786,POLKADOT RAIN HAT,4,2010-12-01 14:32:00,0.85,<NA>,United Kingdom
1446,536544,21787,RAIN PONCHO RETROSPOT,2,2010-12-01 14:32:00,1.66,<NA>,United Kingdom
1447,536544,21790,VINTAGE SNAP CARDS,9,2010-12-01 14:32:00,1.66,<NA>,United Kingdom
1448,536544,21791,VINTAGE HEADS AND TAILS CARD GAME,2,2010-12-01 14:32:00,2.51,<NA>,United Kingdom
1449,536544,21801,CHRISTMAS TREE DECORATION WITH BELL,10,2010-12-01 14:32:00,0.43,<NA>,United Kingdom
1450,536544,21802,CHRISTMAS TREE HEART DECORATION,9,2010-12-01 14:32:00,0.43,<NA>,United Kingdom
1451,536544,21803,CHRISTMAS TREE STAR DECORATION,11,2010-12-01 14:32:00,0.43,<NA>,United Kingdom
1452,536544,21809,CHRISTMAS HANGING TREE WITH BELL,1,2010-12-01 14:32:00,2.51,<NA>,United Kingdom


In [ ]:
#ahora si eliminemos esos registros que se ve que no afecta que podamos eliminarlos
df_retail_2[df_retail_2['id_cliente'].isna()].head(10)

,num_factura,cod_articulo,descripcion,cantidad,fh_factura,precio_initario,id_cliente,pais
1443,536544,21773,DECORATIVE ROSE BATHROOM BOTTLE,1,2010-12-01 14:32:00,2.51,<NA>,United Kingdom
1444,536544,21774,DECORATIVE CATS BATHROOM BOTTLE,2,2010-12-01 14:32:00,2.51,<NA>,United Kingdom
1445,536544,21786,POLKADOT RAIN HAT,4,2010-12-01 14:32:00,0.85,<NA>,United Kingdom
1446,536544,21787,RAIN PONCHO RETROSPOT,2,2010-12-01 14:32:00,1.66,<NA>,United Kingdom
1447,536544,21790,VINTAGE SNAP CARDS,9,2010-12-01 14:32:00,1.66,<NA>,United Kingdom
1448,536544,21791,VINTAGE HEADS AND TAILS CARD GAME,2,2010-12-01 14:32:00,2.51,<NA>,United Kingdom
1449,536544,21801,CHRISTMAS TREE DECORATION WITH BELL,10,2010-12-01 14:32:00,0.43,<NA>,United Kingdom
1450,536544,21802,CHRISTMAS TREE HEART DECORATION,9,2010-12-01 14:32:00,0.43,<NA>,United Kingdom
1451,536544,21803,CHRISTMAS TREE STAR DECORATION,11,2010-12-01 14:32:00,0.43,<NA>,United Kingdom
1452,536544,21809,CHRISTMAS HANGING TREE WITH BELL,1,2010-12-01 14:32:00,2.51,<NA>,United Kingdom


In [ ]:
# Calcular el número de valores ausentes
print(df_retail_2.isna().sum())

num_factura             0
cod_articulo            0
descripcion             0
cantidad                0
fh_factura              0
precio_initario         0
id_cliente         133583
pais                    0
dtype: int64


In [ ]:
df_retail_2[df_retail_2['descripcion'].isna()].head(10)

,num_factura,cod_articulo,descripcion,cantidad,fh_factura,precio_initario,id_cliente,pais
622,536414,22139,NaN,56,2010-12-01 11:52:00,0.0,<NA>,United Kingdom
1970,536545,21134,NaN,1,2010-12-01 14:32:00,0.0,<NA>,United Kingdom
1971,536546,22145,NaN,1,2010-12-01 14:33:00,0.0,<NA>,United Kingdom
1972,536547,37509,NaN,1,2010-12-01 14:33:00,0.0,<NA>,United Kingdom
1987,536549,85226A,NaN,1,2010-12-01 14:34:00,0.0,<NA>,United Kingdom
1988,536550,85044,NaN,1,2010-12-01 14:34:00,0.0,<NA>,United Kingdom
2024,536552,20950,NaN,1,2010-12-01 14:34:00,0.0,<NA>,United Kingdom
2025,536553,37461,NaN,3,2010-12-01 14:35:00,0.0,<NA>,United Kingdom
2026,536554,84670,NaN,23,2010-12-01 14:35:00,0.0,<NA>,United Kingdom
2406,536589,21777,NaN,-10,2010-12-01 16:50:00,0.0,<NA>,United Kingdom


In [ ]:
print(df_retail_2.isna().sum())

num_factura             0
cod_articulo            0
descripcion          1454
cantidad                0
fh_factura              0
precio_initario         0
id_cliente         135037
pais                    0
dtype: int64


In [ ]:
# hagamos un respaldo antes 
df_retail_2 = df_retail

#ahora si eliminemos esos registros que se ve que no afecta que podamos eliminarlos
df_retail_2[df_retail_2['descripcion'].isna()].head(10)

In [ ]:

print('*' * 80)
print('cuenta con valores null--->', df_retail.isnull().sum())
print('*' * 80)
print('*' * 80)

Etapa 2. Preprocesamiento de datos

In [ ]:





print('*' * 80)
print('cuenta con valores null--->', df_retail.isnull().sum())
print('*' * 80)
print('*' * 80)



df_retail['id_cliente'] = df_retail['id_cliente'].astype('string')
df_retail['pais'] = df_retail['pais'].astype('string')



print('*' * 80)
print(df_retail.info())
print('*' * 80)
print('*' * 40 +'df_retail------- revisión de valores ausentes' + '*' * 40)
print(df_retail.isna().sum())




'''

# Bucle en los encabezados reemplazando los valores ausentes con 'unknown'
#columnas a tratar
colum_cambiar =['track', 'artist','genre']
for col in df.columns:
    if col in colum_cambiar:
        df[col]=df[col].fillna('unknown', inplace =False)
print(df.isna().sum())


'''

ausentes = pd.DataFrame({
    'cantidad': df_retail.isna().sum(),
    'porcentaje': (df_retail.isna().sum() / len(df_retail) * 100).round(2),
    'porcentaje %': (((df_retail.isna().sum() / len(df_retail)) * 100).round(2))*100

})

print(ausentes)

#se observa que hay muy un numero muy mayor de id de cliente ausentes, por lo que haremos procedemos a quitarlos
#print(df_retail.isna.head())
print(df_retail[df_retail['id_cliente'].isna()].head(10))

#para saber cuantos son los registros con <NA>
print(df_retail['id_cliente'].isna().sum())

# veamos los id_cliente de que pasi son :

print(
    df_retail[df_retail['id_cliente'].isna()]['pais'].value_counts()
)
'''

# Bucle en los encabezados reemplazando los valores ausentes con 'unknown'
#columnas a tratar
colum_cambiar =['descripcion', 'id_cliente']
for col in df_retail.columns:
    if col in colum_cambiar:
        df_retail[col]=df_retail[col].fillna('unknown', inplace =False)
print(df_retail.isna().sum())



df_retail


'''

********************************************************************************
****************************************df_retail------- solo los 5 registros****************************************
  num_factura cod_articulo                          descripcion  cantidad  \
0      536365       85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1      536365        71053                  WHITE METAL LANTERN         6   
2      536365       84406B       CREAM CUPID HEARTS COAT HANGER         8   
3      536365       84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4      536365       84029E       RED WOOLLY HOTTIE WHITE HEART.         6   

           fh_factura  precio_initario  id_cliente            pais  
0 2010-12-01 08:26:00             2.55     17850.0  United Kingdom  
1 2010-12-01 08:26:00             3.39     17850.0  United Kingdom  
2 2010-12-01 08:26:00             2.75     17850.0  United Kingdom  
3 2010-12-01 08:26:00             3.39     17850.0  United Kin

"\n\n# Bucle en los encabezados reemplazando los valores ausentes con 'unknown'\n#columnas a tratar\ncolum_cambiar =['descripcion', 'id_cliente']\nfor col in df_retail.columns:\n    if col in colum_cambiar:\n        df_retail[col]=df_retail[col].fillna('unknown', inplace =False)\nprint(df_retail.isna().sum())\n\n\n\ndf_retail\n\n\n"